### PR TITLE
cassandra 5.0 vector type CREATE/INSERT support

### PR DESCRIPTION
**Pull request by rukai**
makes progress towards: https://github.com/scylladb/scylla-rust-driver/issues/1014

The vector type is introduced by the currently in beta cassandra 5.
See: https://cassandra.apache.org/doc/latest/cassandra/reference/vector-data-type.html
Scylla does not support vector types and so the tests are setup to only compile/run with a new cassandra_tests config.

This commit does not add support for retrieving the data via a SELECT.
That was omitted to reduce scope and will be implemented in follow up work.

I have split off support for CREATE/INSERT since I value the early feedback of small PRs.
I didnt realize that @pkolaczk had shown interest in implementing this until after I'd went to submit my PR.
Hopefully I've not stepped on any toes.

The cassandra docker-compose.yaml is changed to 5.0 beta cassandra, if this is undesirable I can introduce a second docker-compose.yaml specifically for 5.0

image
Confirmed that the test is running in CI

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/transport/topology.rs
Comment (wprzytula) [inline scylla/src/transport/topology.rs:1630]: The error message is slightly misleading. After all, `4.0` is a valid number, so `.` is a numeric character, but it's never a part of an integer.

File: scylla/src/transport/topology.rs
Comment (wprzytula) [inline scylla/src/transport/topology.rs:190]: Why did you choose specifically `u16` for size? Can you share your reference documentation?

File: test/cluster/cassandra/docker-compose.yml
Comment (wprzytula) [inline test/cluster/cassandra/docker-compose.yml:14]: I'm unsure if we are OK with testing only with newest beta cassandra instead of stable. WDYT @Lorak-mmk?

File: scylla/src/transport/session_test.rs
Comment (wprzytula) [inline scylla/src/transport/session_test.rs:2917]: Important: missing test for a prepared statement combined with Vector.

Comment (wprzytula) [reply scylla/src/transport/session_test.rs:2917]: Question: is Vector insertable as bind value?
I.e., is the following query valid?
`INSERT INTO {}.t (a, b, c) VALUES (1, ?, ?)`
If so, `ColumnType::Vector` variant must be added and ser/de for it must be supported.

**Review by wprzytula — COMMENTED** [review]
What I see, looks good.

However, what I see in this PR is support for parsing schema containing Vector type. What is completely missing is serialization and deserialization support (the latter, you said, you leave for a follow-up). See https://github.com/scylladb/scylla-rust-driver/issues/1014#issuecomment-2178186151 for hints.

File: scylla/src/transport/topology.rs
Comment (rukai) [reply scylla/src/transport/topology.rs]: Good point, Ive now added the reference as a comment above that line.

File: scylla/src/transport/session_test.rs
Comment (rukai) [reply scylla/src/transport/session_test.rs:2917]: I added a test using prepared statements.

I've added bind values to the TODO at the bottom of the test. Lets address them in the follow up work, since when we add `ColumnType::Vector` we will need to add support for all the other vector functionality at the same time.

File: scylla/src/transport/topology.rs
Comment (wprzytula) [inline scylla/src/transport/topology.rs:191]: Now I can see that this is the size limit of a Vector, but what interests me is how Cassandra represents Vector in the system tables. It surely does not use unsigned int, because AFAIK unsigned numbers are not supported by neither ScyllaDB nor Cassandra.
The idea of `PreCqlType` in `topology.rs` is to represent the raw type definition read from system tables. Only later can we convert it to more convenient representation, involving unsigned ints.

Comment (wprzytula) [reply scylla/src/transport/topology.rs:191]: I believe it makes more sense to store size in the same type as Cassandra does.

Comment (rukai) [reply scylla/src/transport/topology.rs:191]: The java driver uses an int, so I'll do the same: https://github.com/apache/cassandra-java-driver/blob/85bb4065098b887d2dda26eb14423ce4fc687045/core/src/main/java/com/datastax/oss/driver/api/core/type/DataTypes.java#L77

Comment (rukai) [reply scylla/src/transport/topology.rs]: I adjusted the code to use an i32 to match the java driver.
I also renamed the `size` to `dimensions` as that is a more accurate name.

File: scylla/src/transport/session_test.rs
Comment (rukai) [reply scylla/src/transport/session_test.rs]: I have changed query string to bind the partition key.

I was already executing the statement, but I've adjusted the execute to provide the binded value.

File: scylla/src/utils/parse.rs
Comment (wprzytula) [inline scylla/src/utils/parse.rs:97]: This has a minor flaw of accepting strings such as `00-21-37`, but such strings will be rejected below anyway, so no problem here.

File: scylla/src/utils/parse.rs
Comment (wprzytula) [inline scylla/src/utils/parse.rs:95]: The the -> The

File: scylla/src/transport/session_test.rs
Comment (wprzytula) [inline scylla/src/transport/session_test.rs:2929]: Could you add a test for schema metadata? Something like this (probably a bit more checks here):
```rust
    let metadata = session.get_cluster_data();
    let columns = &metadata.keyspaces[&ks].tables["t"].columns;
    assert_eq!(
        columns["b"].type_,
        CqlType::Vector(...);
    );
```

Comment (rukai) [reply scylla/src/transport/session_test.rs:2945]: good idea! done.

File: scylla/src/utils/parse.rs
Comment (rukai) [reply scylla/src/utils/parse.rs]: done

**Review by wprzytula — DISMISSED** [review]

File: test/cluster/cassandra/docker-compose.yml
Comment (Lorak-mmk) [reply test/cluster/cassandra/docker-compose.yml:14]: We can use this in addition to stable, but not instead of stable

Comment (rukai) [reply test/cluster/cassandra/docker-compose.yml]: I've moved cassandra-5.0-beta1 into a separate compose file.

File: scylla/src/transport/topology.rs
Comment (Lorak-mmk) [reply scylla/src/transport/topology.rs]: Unresolving this comment because we talked about that with @wprzytula 
I'm not aware of any case were size of vector is stored in some table (as a cql type). IIUC we always parse type names from strings - so the argument about system tables doesn't apply.
Is it possible to create vector with negative length? I'd be very surprised if it was possible. 
If it is not we can and should use unsigned integer for the size (and simplify the parsing function thanks to that).

Should it be u16? I'm not sure. If documentation says that size is limited to 8192 then yes, but I can't find this information is the link in the comment. Could you link the source?

File: .github/workflows/cassandra.yml
Comment (Lorak-mmk) [inline .github/workflows/cassandra.yml:58]: This will be done sequentially - first we perform tests on normal cassandra, then on beta.
Cassandra tests are already longest part of our CI. Maybe it would be better to create a separate workflow for this?

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Thanks for the contribution! It looks good, I left a few minor comments (and responded to a past discussion about a type of vector size).

I believe that the PR title and description are a bit imprecise.

The PR does not add CREATE or INSERT support for Vector type. CREATE is and always was possible, so was INSERT with hardcoded values.

The PR enables the driver to fetch and represent metadata containing Vector type, which is required to fully connect to cluster (because if the driver can't fetch metadata then it also won't fetch peer list).

File: scylla/src/transport/topology.rs
Comment (rukai) [reply scylla/src/transport/topology.rs]: The limit of 2^13 = 8192 is taken from https://cassandra.apache.org/doc/latest/cassandra/reference/vector-data-type.html

However it can be observed that the documentation is incorrect.

From functionally testing, the highest value accepted when creating a table is `2**31 -1 = 2147483647`.
The reason would be that its the highest value representable by a signed 32 bit integer
From functional testing it can also be seen that negative numbers and `0` are rejected.

Do you want me to swap the value from an i32 to a u32?

File: scylla/src/transport/session_test.rs
Comment (rukai) [reply scylla/src/transport/session_test.rs]: I have split up the tests as requested.

File: .github/workflows/cassandra.yml
Comment (rukai) [reply .github/workflows/cassandra.yml]: I have moved it into a separate workflow as requested.

**Comment by rukai** [comment]
> I believe that the PR title and description are a bit imprecise.
>
> The PR does not add CREATE or INSERT support for Vector type. CREATE is and always was possible, so was INSERT with hardcoded values.
>
> The PR enables the driver to fetch and represent metadata containing Vector type, which is required to fully connect to cluster (because if the driver can't fetch metadata then it also won't fetch peer list).

I'm not quite following what you want.
To save both of us the trouble, can you please provide the exact commit message that you want.

**Comment by rukai** [comment]
Hi, any news on this one?

**Comment by wprzytula** [comment]
> Hi, any news on this one?

Sorry for the delay, both @Lorak-mmk and I have been on vacation. Now let @Lorak-mmk take a fresh look.

**Review by Lorak-mmk — COMMENTED** [review]
The PR looks much better now. I am however reluctant to merge it before Cassandra version with Vector support is officially released. Mostly because:
- implementation could change before release
- cql specification seems incomplete w.r.t vectors. They are mentioned in section 5.2 of protocol v5 docs, so we know their serialized form. I don't however see anything about them in section 4.2.5.2 which describes prepared metadata.

Regarding the above: I totally forgot that prepared metadata is a place where we parse the type not from string but from binary form, sorry about that :(
This is also a reason why I favor a bit larger PRs - this would come up when trying to add support for selecting vectors using a prepared statement.

@wprzytula because spec is incomplete, and type used for size depends on it, I think we should wait until C* releases the feature. wdyt?

**Comment by rukai** [comment]
cassandra 5 is released, any interest in landing this?

**Comment by rukai** [comment]
oh right, theres some TODO's in here for when cassandra 5 releases. I'll address those

**Comment by rukai** [comment]
The TODO's have been addressed: the cassandra 5 specific workflow and docker-compose.yaml have been removed, instead we just use the regular cassandra workflow (which is automatically version 5 now since it does not specify a version)

**Comment by rukai** [comment]
oh right, a recent rust release has disallowed that kind of feature check, I'll look into alternatives

**Comment by rukai** [comment]
I applied the same fix as https://github.com/scylladb/scylla-rust-driver/pull/1049 and CI is green now

**Review by wprzytula — DISMISSED** [review]

File: scylla/src/utils/parse.rs
Comment (Lorak-mmk) [inline scylla/src/utils/parse.rs:102]: Two minor issues:
- I don't think we expect integer of LENGTH 2**32 ;) The message should be something like "Expected 32-bit signed integer".
- Looking at other usages and Display definition, I think this should use `ParseErrorCause::Other`.

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Left one minor comment  - after you fix it we can merge I think, and fix potential issue when introducing ser/deser

File: scylla/src/utils/parse.rs
Comment (rukai) [reply scylla/src/utils/parse.rs]: Thankyou for the review, I have addressed your feedback.

**Review by Lorak-mmk — APPROVED** [review]

**Pull request merged**
